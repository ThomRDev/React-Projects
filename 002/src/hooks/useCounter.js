import { useState } from "react"
const useCounter = (initialState) =>{
    const [counter, setCounter] = useState(initialState);
    const increment = (factor = 1) => setCounter(counter + factor)
    const decrement = (factor = 1) => {
        if(counter >= 1) setCounter(counter - factor) 
    }
    const reset = () => setCounter(initialState)


    return {
        counter,
        actions : {
            increment,
            decrement,
            reset
        }
    }

}

export default useCounter