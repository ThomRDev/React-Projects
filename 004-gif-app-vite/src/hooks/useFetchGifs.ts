import { useEffect, useState } from "react"
import { getGifs, IGif} from "../helpers/getGif"

export const useFetchGifs = (value : string)=>{
    const [state,setState] = useState({ data : [] as IGif[], loading : true})
    useEffect(()=>{
        getGifs(value)
        .then(data=>{
            // data?.length &&
            setTimeout(()=>{
                setState({
                    data,
                    loading:false
                })
            },1000)
        })
    },[value])
    return state
}

