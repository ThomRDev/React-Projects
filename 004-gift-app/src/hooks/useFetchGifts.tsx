import { useEffect, useState } from "react"
import { getGif } from "../helpers/getGif"
// export interface DataHook{
//     data : any[],
//     loading : boolean
// }

export const useFetchGifts = (value : string)=>{
    const [state,setState] = useState({ data : [] as object[], loading : true})
    useEffect(()=>{
        getGif(value)
        .then(data=>{
            // data?.length &&
            setTimeout(()=>{
                setState({
                    data,
                    loading:false
                })
            },1000)
        })
        // eslint-disable-next-line
    },[value])
    return state
}

