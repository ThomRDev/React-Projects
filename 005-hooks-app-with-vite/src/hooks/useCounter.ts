import { useState } from "react"

export const useCounter = (initialValue = 0) => {
  const [state,setState] = useState(initialValue)
  const increment = (factor = 1) => setState(state + factor) 
  const subtract = (factor = 1) => setState(state - factor) 
  const reset = () => setState(initialValue)
  return {
    state,
    increment,
    subtract,
    reset
  } 
}