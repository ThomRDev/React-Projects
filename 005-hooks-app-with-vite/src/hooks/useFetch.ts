import { useEffect, useRef, useState } from "react"

const getData = async (url:string) => {
  const response = await fetch(url)
  const data = await response.json()
  return data 
}


export const useFetch = (url:string) => {
  const ref = useRef(true)

  const [state,setState] = useState({
    // cambiar esto de any
    data : [] as any,
    isLoading : true,
    hasError : false
  })

  useEffect(()=>{
    ref.current = true
    setState({
      data : [],
      isLoading: true,
      hasError : false
    })
    getData(url)
    .then((data) => {
      if(ref.current){
        setState({
          data,
          isLoading: false,
          hasError : false
        })
      }
    }).catch((err) => {
      if(ref.current){
        setState({
          data : [],
          isLoading: false,
          hasError : true
        })
      }
    });

    return () => {
      ref.current = false
    }
  },[url])

  return state
}
