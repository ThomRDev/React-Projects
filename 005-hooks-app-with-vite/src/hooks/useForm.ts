import { ChangeEvent, useState } from "react"

// export function useForm <T>(intialValues : T){
export const useForm = <T extends Object>(intialValues : T) => {
  const [form,setForm] = useState(intialValues)
  
  const onChange = (event:ChangeEvent<HTMLInputElement>) => {
    const {name,value} = event.target
    setForm({
      ...form,
      [name] : value
    })
  }

  const onResetForm = () => {
    setForm(intialValues)
  }

  return {
    form,
    onChange,
    onResetForm
  }
}