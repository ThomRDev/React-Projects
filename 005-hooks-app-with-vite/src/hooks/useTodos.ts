import { useEffect, useReducer } from "react"

export type TodoType = {
  id: number
  description : string
  done : boolean
}

const initialState : TodoType[] = [
  {
    id: 1,
    description : "A",
    done : false
  }
]

const todoReducer = (state:TodoType[],action:any):TodoType[] => {
  switch(action.type){
    case "[TODO] ADD TODO":
      return [...state,action.payload]
    case "[TODO] DELETE TODO":
      return state.filter(todo=>todo.id !== action.payload)
    case "[TODO] TOGGLE TODO":
      return state.map(todo=>todo.id === action.payload ? {...todo,done:!todo.done}:todo)
    default:
      return state
  }
}
const init = () => {
  return JSON.parse(localStorage.getItem("todos") as string) || initialState
}

export const useTodos = () => {
  const [state, dispatch] = useReducer(todoReducer, initialState,init)

  const handleNewTodo = (todo:TodoType) => {
    dispatch({
      type :"[TODO] ADD TODO",
      payload : todo
    })
  }
  
  const handleDeleteTodo = (id:number) => {
    dispatch({
      type :"[TODO] DELETE TODO",
      payload : id
    })
  }
  const handleToggleTodo = (id:number) => {
    dispatch({
      type :"[TODO] TOGGLE TODO",
      payload : id
    })
  }
  useEffect(()=>{
    localStorage.setItem("todos",JSON.stringify(state))
  },[state])
  return {
    state,
    todosCount : state.length,
    pendingTodos : state.filter(todo=>!todo.done).length,
    handleDeleteTodo,
    handleNewTodo,
    handleToggleTodo
  }
}