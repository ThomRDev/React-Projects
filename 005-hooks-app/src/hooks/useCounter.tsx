import { useState } from "react"

export const useCounter = (initialState = 0) =>{
    const [counter,setCounter] = useState(initialState)
    
    const decrement = (factor = 1)=>{
        setCounter(counter - factor)
    }
    const increment = (factor = 1)=>{
        setCounter(counter + factor)
    }
    const reset = ()=>setCounter(initialState)
    
    return {
        counter,
        actions:{
            "decrement" : decrement,
            "increment" : increment,
            "reset" : reset,
        }
    }
}