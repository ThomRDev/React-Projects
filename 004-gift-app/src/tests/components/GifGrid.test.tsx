import { shallow } from "enzyme";
import path from "path";
import GiftGrid from "../../components/GiftGrid";
import { useFetchGifts } from "../../hooks/useFetchGifts";


// lo que hace esto es falsear cualquier llamada a este archivo al path, fingir
// controlar la informacion que va a responder
// tendremos que hacer esto si tenemos directamente el
// el custom hook en el component
// si tenemos otro component que contiene a este componente (GiftGrid) ya no es necesario el uso de mocks
jest.mock("../../hooks/useFetchGifts")

// Crear una referencia del custom Hook, pero asignándole otro
// tipo para tener disponible el método "mockReturnValue"
const mockUseFetchGifs = useFetchGifts as jest.MockedFunction<typeof useFetchGifts>;

describe(`Tests in ${path.basename(__filename)} file`, () => {
    
    // without gifs
    test('Must render the component <GifGrid /> correctly - without gifs', () => {

        mockUseFetchGifs.mockReturnValue({
            data  :[],
            loading : true
        })
        // este componete necesitara el custom hooks por eso la linea anterior es falsear la data
        let wrapper = shallow(<GiftGrid value="" />)
        // ya que no existe o no hay data
        expect(wrapper.find("h1").exists()).toBe(false)
        expect(wrapper).toMatchSnapshot()
    })
    // with gif
    test('Must render the component <GifGrid /> correctly with gifs - useFetcgGif', () => {
        mockUseFetchGifs.mockReturnValue({
            data :[
                {
                    id: 'A',
                    title: 'A - By DJ Amuro',
                    url: 'http://localhost/gif/a.gif'
                },
                {
                    id: 'AA',
                    title: 'AA - By DJ Amuro',
                    url: 'http://localhost/gif/aa.gif'
                },
                {
                    id: 'AAA',
                    title: 'AAA - By DJ Amuro',
                    url: 'http://localhost/gif/aaa.gif'
                },
            ],
            loading : false
        })
        let wrapper = shallow(<GiftGrid value="the beatles" />)
        expect(wrapper).toMatchSnapshot()
        expect(wrapper.find("h1").exists()).toBe(true)
        expect( wrapper.find('GiftItem').length).toBe(3);
    })

})