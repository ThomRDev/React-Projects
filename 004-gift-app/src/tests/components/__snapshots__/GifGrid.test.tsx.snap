// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Tests in GifGrid.test.tsx file Must render the component <GifGrid /> correctly - without gifs 1`] = `
<div
  style={
    Object {
      "margin": "4em 0",
      "outline": "1px solid #0D161B",
      "padding": "2em",
      "position": "relative",
    }
  }
>
  <div
    style={
      Object {
        "alignItems": "center",
        "display": "flex",
        "flexDirection": "column",
        "gap": "1em",
        "justifyContent": "center",
      }
    }
  >
    <p
      style={
        Object {
          "color": "#ccc",
          "fontFamily": "monospace",
          "fontSize": "2em",
          "margin": 0,
        }
      }
    >
      Loading
    </p>
    <span
      className="loader"
    />
  </div>
</div>
`;

exports[`Tests in GifGrid.test.tsx file Must render the component <GifGrid /> correctly with gifs - useFetcgGif 1`] = `
<div
  style={
    Object {
      "margin": "4em 0",
      "outline": "1px solid #0D161B",
      "padding": "2em",
      "position": "relative",
    }
  }
>
  <h1
    style={
      Object {
        "color": "#ccc",
        "fontFamily": "monospace",
        "left": "50%",
        "margin": "0",
        "position": "absolute",
        "textTransform": "uppercase",
        "top": "0",
        "transform": "translate(-50%,-50%)",
      }
    }
  >
    the beatles
  </h1>
  <div>
    <GiftItem
      id="A"
      key="A"
      title="A - By DJ Amuro"
      url="http://localhost/gif/a.gif"
    />
    <GiftItem
      id="AA"
      key="AA"
      title="AA - By DJ Amuro"
      url="http://localhost/gif/aa.gif"
    />
    <GiftItem
      id="AAA"
      key="AAA"
      title="AAA - By DJ Amuro"
      url="http://localhost/gif/aaa.gif"
    />
  </div>
</div>
`;
